File: tts.js
#!/usr/bin/env node
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import OpenAI from 'openai';
import dotenv from 'dotenv';
import { getEncoding } from 'js-tiktoken';
import { spawn } from 'child_process';
import { Command } from 'commander';
import cliProgress from 'cli-progress';
import pLimit from 'p-limit';

dotenv.config();

// Determine __dirname equivalent in ESM
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class TTSConfig {
    constructor(cliOptions = {}) {
        this.apiKey = process.env.OPENAI_API_KEY;

        // Load config file if exists
        const configPath = path.join(process.cwd(), 'tts-config.json');
        let fileConfig = {};
        if (fs.existsSync(configPath)) {
            try {
                fileConfig = JSON.parse(fs.readFileSync(configPath, 'utf8'));
            } catch (e) {
                console.warn("⚠️ Failed to parse tts-config.json");
            }
        }

        // Priority: CLI > File > Defaults
        this.model = cliOptions.model || fileConfig.model || "gpt-4o-mini-tts";
        this.voice = cliOptions.voice || fileConfig.voice || "alloy";
        this.responseFormat = cliOptions.format || fileConfig.responseFormat || "mp3";
        this.tokenLimit = parseInt(cliOptions.tokenLimit || fileConfig.tokenLimit || 1600);
        this.maxRetries = parseInt(cliOptions.retries || fileConfig.maxRetries || 3);
        this.retryDelay = 2000;
        this.maxFileSize = 100 * 1024 * 1024;
        this.parallelLimit = parseInt(cliOptions.parallel || fileConfig.parallel || 5);
        this.speed = parseFloat(cliOptions.speed || fileConfig.speed || 1.0);

        if (!this.apiKey) {
            throw new Error("OPENAI_API_KEY environment variable is not set");
        }
    }
}

class FileValidator {
    static validatePath(filePath) {
        try {
            const resolvedPath = path.resolve(filePath);
            const homeDir = process.env.HOME || process.env.USERPROFILE;

            if (!resolvedPath.startsWith(homeDir)) {
                 return { valid: false, message: `File must be located under home directory (${homeDir})` };
            }

            if (!fs.existsSync(resolvedPath)) {
                return { valid: false, message: "File does not exist" };
            }

            const stats = fs.statSync(resolvedPath);
            if (!stats.isFile()) {
                return { valid: false, message: "Specified path is not a file" };
            }

            try {
                fs.accessSync(resolvedPath, fs.constants.R_OK);
            } catch (e) {
                return { valid: false, message: "No read permission for the file" };
            }

            return { valid: true, message: "File path is safe" };

        } catch (e) {
            return { valid: false, message: `File path validation error: ${e.message}` };
        }
    }

    static validateFileSize(filePath, maxSize) {
        try {
            const stats = fs.statSync(filePath);
            if (stats.size > maxSize) {
                return { valid: false, message: `File size too large (limit: ${Math.floor(maxSize/1024/1024)}MB)` };
            }
            return { valid: true, message: "File size OK" };
        } catch (e) {
            return { valid: false, message: `File size check error: ${e.message}` };
        }
    }

    static readFile(filePath) {
        try {
            const content = fs.readFileSync(filePath, 'utf8');
            if (!content.trim()) {
                return { success: false, content: "", message: "File is empty" };
            }
            return { success: true, content: content.trim(), message: "File read successfully" };
        } catch (e) {
            return { success: false, content: "", message: `File reading error: ${e.message}` };
        }
    }
}

class TextProcessor {
    constructor(tokenLimit) {
        this.tokenLimit = tokenLimit;
        this.encoder = getEncoding("cl100k_base");
    }

    splitByTokens(text) {
        const parts = [];
        let current = "";
        const sentences = text.split(/(?<=[。．！？\n])/);

        for (const sentence of sentences) {
             if (!sentence.trim()) continue;

             const testText = current + sentence;
             const tokens = this.encoder.encode(testText);

             if (tokens.length > this.tokenLimit) {
                 if (current.trim()) {
                     parts.push(current.trim());
                 }
                 current = sentence;
             } else {
                 current = testText;
             }
        }

        if (current.trim()) {
            parts.push(current.trim());
        }

        return parts;
    }

    countTokens(text) {
        return this.encoder.encode(text).length;
    }
}

class AudioSynthesizer {
    constructor(config) {
        this.config = config;
        this.openai = new OpenAI({ apiKey: config.apiKey });
    }

    async _createAudio(text) {
        const mp3 = await this.openai.audio.speech.create({
            model: this.config.model,
            voice: this.config.voice,
            input: text,
            speed: this.config.speed,
            response_format: this.config.responseFormat,
        });

        return Buffer.from(await mp3.arrayBuffer());
    }

    async synthesizeWithRetry(text) {
        let lastError = null;

        for (let attempt = 0; attempt < this.config.maxRetries; attempt++) {
            try {
                return await this._createAudio(text);
            } catch (e) {
                lastError = e;
                if (attempt < this.config.maxRetries - 1) {
                    // Silent retry logic to avoid spamming console with parallel requests
                    // Can verify via debug logs if needed
                    await new Promise(resolve => setTimeout(resolve, this.config.retryDelay * (attempt + 1)));
                }
            }
        }
        throw lastError;
    }

    async createSegment(text, index, prefix, outputDir) {
        const audioContent = await this.synthesizeWithRetry(text);
        const partPath = path.join(outputDir, `${prefix}_part${index + 1}.${this.config.responseFormat}`);
        fs.writeFileSync(partPath, audioContent);
        return partPath;
    }
}

class AudioMerger {
    static async mergeFiles(parts, outputPath, outputDir) {
        const listFile = path.join(outputDir, "concat_list.txt");

        try {
            const fileContent = parts.map(p => `file '${p}'`).join('\n');
            fs.writeFileSync(listFile, fileContent);

            await new Promise((resolve, reject) => {
                const ffmpeg = spawn('ffmpeg', [
                    '-f', 'concat',
                    '-safe', '0',
                    '-i', listFile,
                    '-c', 'copy',
                    outputPath
                ]);

                ffmpeg.on('close', (code) => {
                    if (code === 0) resolve();
                    else reject(new Error(`FFmpeg exited with code ${code}`));
                });

                ffmpeg.on('error', (err) => {
                    reject(err);
                });
            });

        } catch (e) {
            throw new Error(`Merge error: ${e.message}`);
        } finally {
            try {
                if (fs.existsSync(listFile)) fs.unlinkSync(listFile);
            } catch (e) { /* ignore */ }
        }
    }
}

class FileCleanup {
    static removeFiles(filePaths) {
        for (const filePath of filePaths) {
            try {
                if (fs.existsSync(filePath)) fs.unlinkSync(filePath);
            } catch (e) { /* ignore */ }
        }
    }
}

class TTSProcessor {
    constructor(options) {
        this.config = new TTSConfig(options);
        this.textProcessor = new TextProcessor(this.config.tokenLimit);
        this.synthesizer = new AudioSynthesizer(this.config);
    }

    async processFile(inputFile) {
        const { valid, message } = FileValidator.validatePath(inputFile);
        if (!valid) {
            console.error(`❌ File path validation error: ${message}`);
            process.exit(1);
        }

        const inputPath = path.resolve(inputFile);
        const { valid: sizeValid, message: sizeMsg } = FileValidator.validateFileSize(inputPath, this.config.maxFileSize);
        if (!sizeValid) {
            console.error(`❌ ${sizeMsg}`);
            process.exit(1);
        }

        const { success, content, message: readMsg } = FileValidator.readFile(inputPath);
        if (!success) {
            console.error(`❌ ${readMsg}`);
            process.exit(1);
        }

        console.log(`\n⚙️  Processing: ${path.basename(inputPath)}`);
        console.log(`   Model: ${this.config.model} | Voice: ${this.config.voice} | Parallel: ${this.config.parallelLimit}`);

        const textParts = this.textProcessor.splitByTokens(content);
        console.log(`📚 Segments: ${textParts.length} chunks`);

        const prefix = path.basename(inputPath, path.extname(inputPath));
        const outputDir = path.dirname(inputPath);
        const outputParts = new Array(textParts.length);

        // Progress Bar
        const bar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        bar.start(textParts.length, 0);

        const limit = pLimit(this.config.parallelLimit);

        try {
            const tasks = textParts.map((partText, index) => {
                return limit(async () => {
                    try {
                        const partPath = await this.synthesizeWithRetryWrapper(partText, index, prefix, outputDir);
                        outputParts[index] = partPath; // Store in correct order
                        bar.increment();
                        return partPath;
                    } catch (e) {
                        throw new Error(`Segment ${index + 1} failed: ${e.message}`);
                    }
                });
            });

            await Promise.all(tasks);

        } catch (e) {
            bar.stop();
            console.error(`\n❌ Failed to generate audio: ${e.message}`);
            // Cleanup partials if needed, or leave for debug?
            // Cleanup is safer.
            FileCleanup.removeFiles(outputParts.filter(p => p));
            process.exit(1);
        }

        bar.stop();
        console.log("\n🔄 Merging audio segments...");

        const mergedFile = path.join(outputDir, `${prefix}_merged.${this.config.responseFormat}`);

        try {
            await AudioMerger.mergeFiles(outputParts, mergedFile, outputDir);
        } catch (e) {
            console.error(`❌ Failed to merge audio files: ${e.message}`);
            process.exit(1);
        }

        FileCleanup.removeFiles(outputParts);

        console.log(`✅ Done! Saved to: ${mergedFile}\n`);
    }

    async synthesizeWithRetryWrapper(text, index, prefix, outputDir) {
        return this.synthesizer.createSegment(text, index, prefix, outputDir);
    }
}

async function main() {
    const program = new Command();

    program
        .name('tts')
        .description('Convert text file to speech using OpenAI API')
        .version('2.0.0')
        .argument('<file>', 'Input text file path')
        .option('-m, --model <model>', 'OpenAI model', 'gpt-4o-mini-tts')
        .option('-v, --voice <voice>', 'Voice (alloy, echo, fable, onyx, nova, shimmer)', 'alloy')
        .option('-p, --parallel <number>', 'Concurrent request limit', '5')
        .option('-s, --speed <number>', 'Speed (0.25 to 4.0)', '1.0')
        .action(async (file, options) => {
            try {
                const processor = new TTSProcessor(options);
                await processor.processFile(file);
            } catch (e) {
                if (e.message.includes("OPENAI_API_KEY")) {
                    console.error(`❌ Configuration error: ${e.message}`);
                } else {
                    console.error(`❌ Unexpected error: ${e.message}`);
                }
                process.exit(1);
            }
        });

    program.parse(process.argv);
}

main();

File: tts.py
import openai
import os
import sys
import tiktoken
import re
from pathlib import Path
from dotenv import load_dotenv
import subprocess
import time
from typing import List, Tuple, Optional


class TTSConfig:
    """Configuration class for TTS settings"""

    def __init__(self):
        load_dotenv()
        self.api_key = os.getenv("OPENAI_API_KEY")
        self.model = "gpt-4o-mini-tts"
        self.voice = "alloy"
        self.response_format = "mp3"
        self.token_limit = 1600
        self.max_retries = 3
        self.retry_delay = 2
        self.max_file_size = 100 * 1024 * 1024  # 100MB

        self.instructions = (
            "Speak with a warm, clear, and engaging tone suitable for educational narration in both Japanese and English. "
            "Use natural pacing with gentle emphasis on key points. "
            "For Japanese text: pronounce each character clearly with proper pitch accent and natural rhythm. "
            "For English in Japanese context: maintain clear pronunciation while flowing naturally. "
            "Add appropriate pauses at punctuation marks and between sentences for comprehension. "
            "Maintain consistent energy and deliver as a friendly, knowledgeable tutor."
        )

        # Initialize OpenAI API
        if not self.api_key:
            raise ValueError("OPENAI_API_KEY environment variable is not set")
        openai.api_key = self.api_key


class FileValidator:
    """File validation utilities"""

    @staticmethod
    def validate_path(file_path: str) -> Tuple[bool, str]:
        """Validate the safety and accessibility of the file path"""
        try:
            path = Path(file_path).resolve()

            # Prevent path traversal attacks
            if ".." in str(path):
                return False, "Path contains '..' (potential path traversal attack)"

            # Check if absolute path is within safe range
            home_dir = Path.home()
            try:
                path.relative_to(home_dir)
            except ValueError:
                return False, f"File must be located under home directory ({home_dir})"

            # Check file existence and type
            if not path.exists():
                return False, "File does not exist"

            if not path.is_file():
                return False, "Specified path is not a file"

            # Check permissions
            if not os.access(path, os.R_OK):
                return False, "No read permission for the file"

            return True, "File path is safe"

        except Exception as e:
            return False, f"File path validation error: {str(e)}"

    @staticmethod
    def validate_file_size(file_path: Path, max_size: int) -> Tuple[bool, str]:
        """Validate file size"""
        try:
            if file_path.stat().st_size > max_size:
                return False, f"File size too large (limit: {max_size//1024//1024}MB)"
            return True, "File size OK"
        except Exception as e:
            return False, f"File size check error: {str(e)}"

    @staticmethod
    def read_file(file_path: Path) -> Tuple[bool, str, str]:
        """Read and validate file content"""
        try:
            with file_path.open("r", encoding="utf-8") as f:
                content = f.read().strip()

            if not content:
                return False, "", "File is empty"

            return True, content, "File read successfully"

        except UnicodeDecodeError:
            return False, "", "File is not UTF-8 encoded"
        except Exception as e:
            return False, "", f"File reading error: {e}"


class TextProcessor:
    """Text processing utilities"""

    def __init__(self, token_limit: int):
        self.token_limit = token_limit
        self.encoder = tiktoken.get_encoding("cl100k_base")

    def split_by_tokens(self, text: str) -> List[str]:
        """Split text into segments based on token limit"""
        parts = []
        current = ""

        sentences = re.split(r'(?<=[。．！？\n])', text)

        for sentence in sentences:
            if not sentence.strip():
                continue

            test_text = current + sentence
            tokens = self.encoder.encode(test_text)

            if len(tokens) > self.token_limit:
                if current.strip():
                    parts.append(current.strip())
                current = sentence
            else:
                current = test_text

        if current.strip():
            parts.append(current.strip())

        return parts

    def count_tokens(self, text: str) -> int:
        """Count tokens in text"""
        return len(self.encoder.encode(text))


class AudioSynthesizer:
    """Audio synthesis with retry logic"""

    def __init__(self, config: TTSConfig):
        self.config = config

    def _create_audio(self, text: str) -> bytes:
        """Create audio using OpenAI API"""
        response = openai.audio.speech.create(
            model=self.config.model,
            voice=self.config.voice,
            input=text,
            instructions=self.config.instructions,
            response_format=self.config.response_format,
        )
        return response.content

    def synthesize_with_retry(self, text: str) -> bytes:
        """Synthesize audio with retry logic"""
        last_error = None

        for attempt in range(self.config.max_retries):
            try:
                return self._create_audio(text)
            except Exception as e:
                last_error = e
                if attempt < self.config.max_retries - 1:
                    print(f"⚠️ API call failed (attempt {attempt + 1}/{self.config.max_retries}): {str(e)}")
                    print(f"🔄 Retrying in {self.config.retry_delay} seconds...")
                    time.sleep(self.config.retry_delay * (attempt + 1))
                else:
                    print(f"❌ API call failed {self.config.max_retries} times")

        raise last_error

    def create_segment(self, text: str, index: int, prefix: str, output_dir: Path) -> Path:
        """Create a single audio segment"""
        print(f"🎙️ Generating Part {index + 1}...")

        audio_content = self.synthesize_with_retry(text)

        part_path = output_dir / f"{prefix}_part{index + 1}.mp3"
        with open(part_path, "wb") as f:
            f.write(audio_content)

        print(f"✅ Saved: {part_path}")
        return part_path


class AudioMerger:
    """Audio file merging utilities"""

    @staticmethod
    def merge_files(parts: List[Path], output_path: Path, output_dir: Path) -> None:
        """Merge multiple audio files into one"""
        list_file = output_dir / "concat_list.txt"

        try:
            # Create concatenation list file
            with open(list_file, "w") as f:
                for part_path in parts:
                    f.write(f"file '{part_path}'\n")

            print("🔄 Merging audio segments...")

            # Run FFmpeg
            subprocess.run([
                "ffmpeg", "-f", "concat", "-safe", "0", "-i", str(list_file),
                "-c", "copy", str(output_path)
            ], check=True)

            print(f"🎧 Merged file created: {output_path}")

        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"FFmpeg execution error: {e}")
        except Exception as e:
            raise RuntimeError(f"File merge error: {e}")
        finally:
            # Clean up temporary file
            try:
                os.remove(list_file)
            except:
                pass


class FileCleanup:
    """File cleanup utilities"""

    @staticmethod
    def remove_files(file_paths: List[Path]) -> None:
        """Remove multiple files safely"""
        for file_path in file_paths:
            try:
                os.remove(file_path)
                print(f"🗑️ Deleted: {file_path}")
            except Exception as e:
                print(f"⚠️ Failed to delete: {file_path} ({e})")

    @staticmethod
    def cleanup_on_error(file_paths: List[Path]) -> None:
        """Clean up files when an error occurs"""
        for file_path in file_paths:
            try:
                os.remove(file_path)
            except:
                pass


class TTSProcessor:
    """Main TTS processing class"""

    def __init__(self):
        self.config = TTSConfig()
        self.validator = FileValidator()
        self.text_processor = TextProcessor(self.config.token_limit)
        self.synthesizer = AudioSynthesizer(self.config)
        self.merger = AudioMerger()
        self.cleanup = FileCleanup()

    def process_file(self, input_file: str) -> None:
        """Process a text file and convert to audio"""
        # Validate file path
        is_valid, message = self.validator.validate_path(input_file)
        if not is_valid:
            print(f"❌ File path validation error: {message}")
            sys.exit(1)

        print(f"✅ File path validation OK: {message}")

        input_path = Path(input_file).resolve()

        # Validate file size
        is_valid, message = self.validator.validate_file_size(input_path, self.config.max_file_size)
        if not is_valid:
            print(f"❌ {message}")
            sys.exit(1)

        # Read file content
        success, content, message = self.validator.read_file(input_path)
        if not success:
            print(f"❌ {message}")
            sys.exit(1)

        # Process text
        print("✂️ Splitting text based on token length...")
        text_parts = self.text_processor.split_by_tokens(content)
        print(f"📚 Number of segments: {len(text_parts)}")

        # Generate audio segments
        prefix = input_path.stem
        output_dir = input_path.parent
        output_parts = []

        try:
            for i, part_text in enumerate(text_parts):
                char_count = len(part_text)
                token_count = self.text_processor.count_tokens(part_text)
                print(f"🧮 Part {i+1} Character count: {char_count} / Token count: {token_count}")

                part_path = self.synthesizer.create_segment(part_text, i, prefix, output_dir)
                output_parts.append(part_path)

        except Exception as e:
            print(f"❌ Failed to generate audio: {e}")
            self.cleanup.cleanup_on_error(output_parts)
            sys.exit(1)

        # Merge audio files
        merged_file = output_dir / f"{prefix}_merged.mp3"

        try:
            self.merger.merge_files(output_parts, merged_file, output_dir)
        except Exception as e:
            print(f"❌ Failed to merge audio files: {e}")
            sys.exit(1)

        # Clean up temporary files
        self.cleanup.remove_files(output_parts)

        print("✅ All done!")
        print(f"🎧 Output file: {merged_file}")


def main():
    """Main entry point"""
    if len(sys.argv) < 2:
        print("Usage: python tts.py input.txt")
        sys.exit(1)

    try:
        processor = TTSProcessor()
        processor.process_file(sys.argv[1])
    except ValueError as e:
        print(f"❌ Configuration error: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
